flashrom: 1.3.0 -> 1.4.0

edef1c opened the pull request:
## Description of changes

[Release notes](https://flashrom.org/release_notes/v_1_4.html)

The most significant change for us is that future versions won't support Makefile builds, with Meson becoming the only supported build system. I've switched the build to Meson, rather than saving that surprise for the next update.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

edef1c committed:
flashrom: 1.3.0 -> 1.4.0

edef1c committed:
flashrom: build with meson

Per the 1.4.0 release notes, Meson will be the only supported build
system for future versions.

edef1c commented:
Somewhat of an editorial decision: The release notes for 1.4.0 state that Promontory chipsets are unsupported, and that attempting to read flash on those systems may crash the system. The patch that removes this (broken) support only made it into the 1.3.0 release, seemingly by mistake, and the relevant code has been essentially untouched since.

I'd rather not crash the systems of unsuspecting NixOS/nixpkgs users, so I've forward-ported the patch from 1.3.0.
If anyone feels daring enough to re-enable this, they can drop the patch themselves.

edef1c committed:
flashrom: remove (broken) Promontory support

Release notes for 1.4.0 state that Promontory chipsets are unsupported,
and that attempting to read flash on those systems may crash the system.

The patch that removes this (broken) support only made it into the
1.3.0 release, seemingly by mistake, and the relevant code has been
essentially untouched since.

We'd rather not crash the systems of unsuspecting NixOS/nixpkgs users,
so this change forward-ports the patch from 1.3.0, though amended to
reference the relevant bug in the error message, rather than requesting
the user email upstream.

Link: https://ticket.coreboot.org/issues/370
Link: https://review.coreboot.org/c/flashrom/+/68824

edef1c commented:
Slightly amending that patch to give an informative error message referencing the upstream bug, rather than defaulting to the standard "Please report this to flashrom@flashrom.org" message, so that we don't cause unnecessary trouble and confusion for upstream.

fpletz commented:
I was waiting for ofborg to finish the darwin builds before merging this. Unfortunately x86_64-darwin failed:

```
FAILED: tests/flashrom_unit_tests.p/erase_func_algo.c.o 
clang -Itests/flashrom_unit_tests.p -Itests -I../tests -I../include -I/nix/store/cdqpvwakdw8pijxbx5s9q4vzagpzpvyr-cmocka-1.1.7/include -I/nix/store/ixz3hwjcnji6lmc6d1c70j29fqfh7w8w-libusb-1.0.27-dev/include/libusb-1.0 -I/nix/store/aaykkdhkn65kxi37m74krq99x6zc7kk9-libftdi-1.5-unstable-2023-12-21/include/libftdi1 -fdiagnostics-color=always -Wall -Winvalid-pch -Wextra -Werror -std=c99 -Wshadow -Wmissing-prototypes -Wwrite-strings -Wno-unused-parameter -Wno-address-of-packed-member -Wno-enum-conversion -Wno-missing-braces -D_DEFAULT_SOURCE -D_POSIX_C_SOURCE=200809L -D_BSD_SOURCE -D__BSD_VISIBLE -D__XSI_VISIBLE -D_NETBSD_SOURCE -D_DARWIN_C_SOURCE '-DFLASHROM_VERSION="1.4.0"' -DHAVE_CLOCK_GETTIME=1 -DHAVE_STRNLEN=1 -DHAVE_GETOPT_H=1 -DHAVE_UTSNAME=1 -DIS_WINDOWS=0 -D__FLASHROM_LITTLE_ENDIAN__=1 -includestdlib.h -includeunittest_env.h -DCONFIG_DELAY_MINIMUM_SLEEP_US=100000 -DCONFIG_BUSPIRATE_SPI=1 -DCONFIG_CH341A_SPI=1 -DCONFIG_CH347_SPI=1 -DCONFIG_DEDIPROG=1 -DCONFIG_DEVELOPERBOX_SPI=1 -DCONFIG_DIGILENT_SPI=1 -DCONFIG_DIRTYJTAG_SPI=1 -DCONFIG_DUMMY=1 -DCONFIG_FT2232_SPI=1 -DCONFIG_PICKIT2_SPI=1 -DCONFIG_PONY_SPI=1 -DCONFIG_RAIDEN_DEBUG_SPI=1 -DCONFIG_SERPROG=1 -DCONFIG_STLINKV3_SPI=1 -DCONFIG_USBBLASTER_SPI=1 -DCONFIG_DEFAULT_PROGRAMMER_NAME=NULL '-DCONFIG_DEFAULT_PROGRAMMER_ARGS=""' -ffunction-sections -fdata-sections -U_FORTIFY_SOURCE -DFLASHROM_TEST -MD -MQ tests/flashrom_unit_tests.p/erase_func_algo.c.o -MF tests/flashrom_unit_tests.p/erase_func_algo.c.o.d -o tests/flashrom_unit_tests.p/erase_func_algo.c.o -c ../tests/erase_func_algo.c
../tests/erase_func_algo.c:210:4: error: too many arguments provided to function-like macro invocation
                        block_erase_chip_3,     // 4 bytes
                        ^
/nix/store/hmchx96ysd0h19p7kx0w12pljkbp30j3-Libsystem-1238.60.2/include/secure/_string.h:64:9: note: macro 'memcpy' defined here
#define memcpy(dest, src, len)                                  \
        ^
../tests/erase_func_algo.c:206:2: note: parentheses are required around macro argument containing braced initializer list
        memcpy(g_test_erase_injector,
        ^
../tests/erase_func_algo.c:1026:4: error: too many arguments provided to function-like macro invocation
                        }
                        ^
/nix/store/hmchx96ysd0h19p7kx0w12pljkbp30j3-Libsystem-1238.60.2/include/secure/_string.h:64:9: note: macro 'memcpy' defined here
#define memcpy(dest, src, len)                                  \
        ^
../tests/erase_func_algo.c:1020:2: note: parentheses are required around macro argument containing braced initializer list
        memcpy(
        ^
../tests/erase_func_algo.c:1239:4: error: too many arguments provided to function-like macro invocation
                        block_erase_chip_with_protected_region_3,
                        ^
/nix/store/hmchx96ysd0h19p7kx0w12pljkbp30j3-Libsystem-1238.60.2/include/secure/_string.h:64:9: note: macro 'memcpy' defined here
#define memcpy(dest, src, len)                                  \
        ^
../tests/erase_func_algo.c:1235:2: note: parentheses are required around macro argument containing braced initializer list
        memcpy(g_test_erase_injector,
        ^
../tests/erase_func_algo.c:1328:4: error: too many arguments provided to function-like macro invocation
                        block_erase_chip_with_protected_region_3,
                        ^
/nix/store/hmchx96ysd0h19p7kx0w12pljkbp30j3-Libsystem-1238.60.2/include/secure/_string.h:64:9: note: macro 'memcpy' defined here
#define memcpy(dest, src, len)                                  \
        ^
../tests/erase_func_algo.c:1324:2: note: parentheses are required around macro argument containing braced initializer list
        memcpy(g_test_erase_injector,
        ^
4 errors generated.
ninja: build stopped: subcommand failed.
error: builder for '/nix/store/3085k8kdsxsb1cl0ik8azn7l1dcy1gpn-flashrom-1.4.0.drv' failed with exit code 1;
       last 10 log lines:
       >                         block_erase_chip_with_protected_region_3,
       >                         ^
       > /nix/store/hmchx96ysd0h19p7kx0w12pljkbp30j3-Libsystem-1238.60.2/include/secure/_string.h:64:9: note: macro 'memcpy' defined here
       > #define memcpy(dest, src, len)                                  \
       >         ^
       > ../tests/erase_func_algo.c:1324:2: note: parentheses are required around macro argument containing braced initializer list
       >         memcpy(g_test_erase_injector,
       >         ^
       > 4 errors generated.
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/3085k8kdsxsb1cl0ik8azn7l1dcy1gpn-flashrom-1.4.0.drv'.
```

https://logs.ofborg.org/?key=nixos/nixpkgs.330581&attempt_id=2abb57f4-c2ba-4280-a590-673885553f12

Old version builds fine on hydra: https://hydra.nixos.org/job/nixpkgs/trunk/flashrom.x86_64-darwin

edef1c committed:
flashrom: fix Darwin build

memcpy is a macro on Darwin, and thus requires braced initializer lists
to be parenthesized.

edef1c commented:
I've just patched that, but I'll have to wait for ofborg to build it to find out if that resolves everything, since I don't have a Darwin reference machine available.

edef1c commented:
Looks like I didn't catch both of the relevant memcpys, let's give that another spin.

smancill commented:
Still failing on Darwin.

alyssais commented on pkgs/tools/misc/flashrom/default.nix:
```diff
@@ -23,14 +23,17 @@ stdenv.mkDerivation rec {
+      ./0001-sb600spi.c-Drop-Promontory-support.patch
+    ]
+    # Darwin's memcpy is a macro, and requires braced initializer lists to be parenthesized
+    ++ lib.optional stdenv.isDarwin ./0002-tests-erase-parenthesize-braced-initializer.patch;
```
Upstream?

alyssais replied:
Also, we try to avoid conditional patch applications where there isn't a strong reason for them to be conditional, because it makes it harder to notice when they need to be updated.  In this case, there's no harm to other platforms from applying this patch unconditionally.
